streaming/base/ingest_util.py

```diff
@@ -0,0 +1,239 @@
+
+                        sz = beautify(kb)
+                        cnt = observer.file_count
+
+                        if elapsed_time % 10 == 0:
+                            logger.warning(
+                                f'Downloaded {cnt} files, Total approx file size = {sz}, Time Elapsed: {elapsed_time} seconds.'
+                            )
+
+                        if elapsed_time > 0 and elapsed_time % 120 == 0:
+                            observer.get_directory_info(
+                            )  # Get the actual stats by walking through the directory
+                            observer.average_file_size = observer.file_size // observer.file_count
+                            logger.warning(
+                                f'update average file size to {beautify(observer.average_file_size//1024)}'
+                            )
+
+                        time.sleep(1)
+                    except Exception as exc:
+                        # raise RuntimeError("Something bad happened") from exc
+                        logger.warning(str(exc))
+                        time.sleep(1)
+                        continue
+
+            monitor_thread = threading.Thread(target=monitor_directory)
+            monitor_thread.start()
+
+            try:
+                result = func(repo_id, local_dir, max_workers, token, allow_patterns, *args,
+                              **kwargs)
+                return result
+            finally:
+                observer.get_directory_info(
+                )  # Get the actual stats by walking through the directory
+                logger.warning(
+                    f'Done! Downloaded {observer.file_count} files, Total file size = {beautify(observer.file_size//1024)}, Time Elapsed: {int(time.time() - observer.tik)} seconds.'
+                )
+                observer.stop()
+                observer.join()
+
+                event.set()
+                monitor_thread.join()
+
+        return wrapper
+
+    return decorator
+
+
+@monitor_directory_changes()
+@retry([Exception, RuntimeError], num_attempts=10, initial_backoff=10)
+def hf_snapshot(repo_id: str, local_dir: str, max_workers: int, token: str,
+                allow_patterns: Optional[List[str]]):
+    """API call to HF snapshot_download.
+
+    which internally use hf_hub_download
+    """
+    print(
+        f'Now start to download {repo_id} to {local_dir}, with allow_patterns = {allow_patterns}')
+    output = snapshot_download(repo_id,
+                               repo_type='dataset',
+                               local_dir=local_dir,
+                               local_dir_use_symlinks=False,
+                               max_workers=max_workers,
+                               resume_download=True,
+                               token=token,
+                               allow_patterns=allow_patterns)
+    return output
+
+
+def download_hf_dataset(local_cache_directory: str,
```

Comment:
Does this method support downloading all the popular HF Generative AI datasets?

Reply:
This would be a great pr to see if we could abuse the new fsspec `hf` fsspec system to start accessing files. we would just need to use `fsspec open` and the related ls shims to navigate through the directories.